contracts/user/SharedConsumer.sol

```diff
@@ -0,0 +1,314 @@
+    BasicRandcastConsumerBase,
+    UUPSUpgradeable,
+    OwnableUpgradeable
+{
+    // To be update
+    uint32 private constant DRAW_CALLBACK_GAS_BASE = 1200000;
+    uint32 private constant ROLL_CALLBACK_GAS_BASE = 1000000;
+    uint32 private constant FEE_OVERHEAD = 1000000;
+    uint32 private constant GROUP_SIZE = 3;
+    uint32 private constant MAX_GAS_LIMIT = 2000000;
+
+    // common subId for trial
+    uint64 trialSubId;
+
+    mapping(address => uint64) public userSubIds;
+
+    mapping(bytes32 => RequestData) public pendingRequests;
+
+    struct RequestData {
+        PlayType playType;
+        bytes param;
+    }
+
+    enum PlayType {
+        Draw,
+        Roll
+    }
+
+    event RollDiceRequest(
+        address user,
+        uint64 subId,
+        bytes32 requestId,
+        uint32 bunch,
+        uint32 size,
+        uint256 paidAmount,
+        uint256 seed,
+        uint16 requestConfirmations
+    );
+
+    event DrawTicketsRequest(
+        address user,
+        uint64 subId,
+        bytes32 requestId,
+        uint32 totalNumber,
+        uint32 winnerNumber,
+        uint256 paidAmount,
+        uint256 seed,
+        uint16 requestConfirmations
+    );
+
+    event RollDiceResult(bytes32 requestId, uint256[] result);
+    event DrawTicketsResult(bytes32 requestId, uint256[] result);
+
+    error InsufficientFund(uint256 fundAmount, uint256 requiredAmount);
+    error InvalidSubId();
+    error GasLimitTooBig(uint256 have, uint32 want);
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(address adapter) BasicRandcastConsumerBase(adapter) {
+        _disableInitializers();
+    }
+
+    function initialize() public initializer {
+        __Ownable_init(msg.sender);
+    }
+    // solhint-disable-next-line no-empty-blocks
+    function _authorizeUpgrade(address) internal override onlyOwner {}
+
+    function _fundSubId(PlayType playType, uint64 subId, bytes memory params) internal returns (uint64) {
+        if (subId == trialSubId) {
+            return subId;
+        }
+        if (subId == 0) {
+            subId = userSubIds[msg.sender];
+            if (subId == 0) {
+                subId = IAdapter(adapter).createSubscription();
+                IAdapter(adapter).addConsumer(subId, address(this));
+                userSubIds[msg.sender] = subId;
+            }
+        }
+        uint256 fundAmount = estimateFee(playType, subId, params);
+        if (msg.value < fundAmount) {
+            revert InsufficientFund({fundAmount: fundAmount, requiredAmount: msg.value});
+        }
+        if (fundAmount == 0) {
+            return subId;
+        }
+        IAdapter(adapter).fundSubscription{value: msg.value}(subId);
+        return subId;
+    }
+
+    struct Subscription {
+        uint256 balance;
+        uint256 inflightCost;
+        uint64 reqCount;
+        uint64 freeRequestCount;
+        uint64 reqCountInCurrentPeriod;
+        uint256 lastRequestTimestamp;
+    }
```

Comment:
Move this to the top. The layout is like constant variables, public variables, private variables, structs, events, errors, then public/external transactions, views, and internal functions.